### PR TITLE
Support IntPtr and UIntPtr serialization

### LINKED ISSUE #1735: Sentry JsonConverter can't handle IntPtr/UIntPtr/nint/nuint types.
### Package

Sentry

### .NET Flavor

.NET

### .NET Version

6.0.6

### OS

Windows

### SDK Version

3.18.0

### Self-Hosted Sentry Version

_No response_

### Steps to Reproduce

Try to declare a context with a member of type nint, nuint, IntPtr, UIntPtr

### Expected Result

Member would be treated as an int, uint, long or ulong as appropriate for the platform.

### Actual Result

```
System.NotSupportedException: Serialization and deserialization of 'System.UIntPtr' instances are not supported.
   at System.Text.Json.Serialization.Converters.UnsupportedTypeConverter`1.Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
```

<details><summary>Expanded stack trace...</summary>

```
   at System.Text.Json.Serialization.Converters.UnsupportedTypeConverter`1.Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCoreAsObject(Utf8JsonWriter writer, Object value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.JsonSerializer.WriteUsingSerializer[TValue](Utf8JsonWriter writer, TValue& value, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Serialize[TValue](Utf8JsonWriter writer, TValue value, JsonSerializerOptions options)
   at Sentry.Internal.Extensions.JsonExtensions.WriteDynamicValue(Utf8JsonWriter writer, Object value, IDiagnosticLogger logger) in /_/src/Sentry/Internal/Extensions/JsonExtensions.cs:line 337
   at Sentry.Internal.Extensions.JsonExtensions.WriteDynamic(Utf8JsonWriter writer, String propertyName, Object value, IDiagnosticLogger logger) in /_/src/Sentry/Internal/Extensions/JsonExtensions.cs:line 351
   at Sentry.Internal.Extensions.JsonExtensions.WriteDictionaryValue(Utf8JsonWriter writer, IEnumerable`1 dic, IDiagnosticLogger logger) in /_/src/Sentry/Internal/Extensions/JsonExtensions.cs:line 151
   at Sentry.Contexts.WriteTo(Utf8JsonWriter writer, IDiagnosticLogger logger) in /_/src/Sentry/Protocol/Contexts.cs:line 95
   at Sentry.Internal.Extensions.JsonExtensions.WriteSerializableValue(Utf8JsonWriter writer, IJsonSerializable value, IDiagnosticLogger logger) in /_/src/Sentry/Internal/Extensions/JsonExtensions.cs:line 269
   at Sentry.Internal.Extensions.JsonExtensions.WriteSerializable(Utf8JsonWriter writer, String propertyName, IJsonSerializable value, IDiagnosticLogger logger) in /_/src/Sentry/Internal/Extensions/JsonExtensions.cs:line 279
   at Sentry.Internal.Extensions.JsonExtensions.WriteSerializableIfNotNull(Utf8JsonWriter writer, String propertyName, IJsonSerializable value, IDiagnosticLogger logger) in /_/src/Sentry/Internal/Extensions/JsonExtensions.cs:line 515
   at Sentry.SentryEvent.WriteTo(Utf8JsonWriter writer, IDiagnosticLogger logger) in /_/src/Sentry/SentryEvent.cs:line 242
   at Sentry.Protocol.Envelopes.JsonSerializable.SerializeAsync(Stream stream, IDiagnosticLogger logger, CancellationToken cancellationToken) in /_/src/Sentry/Envelopes/JsonSerializable.cs:line 35
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Protocol.Envelopes.JsonSerializable.SerializeAsync(Stream stream, IDiagnosticLogger logger, CancellationToken cancellationToken)
   at Sentry.Protocol.Envelopes.EnvelopeItem.BufferPayloadAsync(IDiagnosticLogger logger, CancellationToken cancellationToken) in /_/src/Sentry/Envelopes/EnvelopeItem.cs:line 88
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Protocol.Envelopes.EnvelopeItem.BufferPayloadAsync(IDiagnosticLogger logger, CancellationToken cancellationToken)
   at Sentry.Protocol.Envelopes.EnvelopeItem.SerializeAsync(Stream stream, IDiagnosticLogger logger, CancellationToken cancellationToken) in /_/src/Sentry/Envelopes/EnvelopeItem.cs:line 156
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Protocol.Envelopes.EnvelopeItem.SerializeAsync(Stream stream, IDiagnosticLogger logger, CancellationToken cancellationToken)
   at Sentry.Protocol.Envelopes.Envelope.SerializeAsync(Stream stream, IDiagnosticLogger logger, CancellationToken cancellationToken) in /_/src/Sentry/Envelopes/Envelope.cs:line 82
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Protocol.Envelopes.Envelope.SerializeAsync(Stream stream, IDiagnosticLogger logger, CancellationToken cancellationToken)
   at Sentry.Internal.Http.EnvelopeHttpContent.SerializeToStreamAsync(Stream stream, TransportContext context) in /_/src/Sentry/Internal/Http/EnvelopeHttpContent.cs:line 30
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Internal.Http.EnvelopeHttpContent.SerializeToStreamAsync(Stream stream, TransportContext context)
   at System.Net.Http.HttpContent.SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.InternalCopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.CopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.CopyToAsync(Stream stream, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.CopyToAsync(Stream stream)
   at Sentry.Internal.Http.GzipBufferedRequestBodyHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /_/src/Sentry/Internal/Http/GzipBufferedRequestBodyHandler.cs:line 63
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Internal.Http.GzipBufferedRequestBodyHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Sentry.Internal.Http.RetryAfterHandler.<>n__0(HttpRequestMessage request, CancellationToken cancellationToken)
   at Sentry.Internal.Http.RetryAfterHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /_/src/Sentry/Internal/Http/RetryAfterHandler.cs:line 65
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Internal.Http.RetryAfterHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageInvoker.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<>n__0(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Sentry.Internal.Http.HttpTransport.SendEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken) in /_/src/Sentry/Internal/Http/HttpTransport.cs:line 46
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Sentry.Internal.Http.HttpTransport.SendEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken)
   at Sentry.Internal.BackgroundWorker.WorkerAsync() in /_/src/Sentry/Internal/BackgroundWorker.cs:line 123
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetExistingTaskResult(Task`1 task, TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutAsync(TaskNode asyncWaiter, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Threading.Tasks.Task.CancellationPromise`1.System.Threading.Tasks.ITaskCompletionAction.Invoke(Task completingTask)
   at System.Threading.Tasks.CompletionActionInvoker.System.Threading.IThreadPoolWorkItem.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```
</details>

exception
call stack

### PR DESCRIPTION
**Pull request by SimonCropp**
fixes #1735

### CODE REVIEW

**Review by bruno-garcia — APPROVED** [review]
LGTM

**Comment by SimonCropp** [comment]
@bruno-garcia one thing of concern is nuinit and nint tests passed with no mods. which seems to conflict with the issue #1735

File: test/Sentry.Tests/SerializationTests.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        using var stream = new MemoryStream();
+        await using (var writer = new Utf8JsonWriter(stream))
+        {
+            writer.WriteDynamicValue(target, null);
+        }
+
+        var json = Encoding.UTF8.GetString(stream.ToArray());
+        await Verify(json).UseParameters(name);
+    }
```
Comment (mattjohnsonpint) [inline test/Sentry.Tests/SerializationTests.cs:18]:
```suggestion
using System.Text.Json;
using Sentry.Testing;

[UsesVerify]
public class SerializationTests
{
    private readonly IDiagnosticLogger _testOutputLogger;
    
    public SerializationTests (ITestOutputHelper output)
    {
        _testOutputLogger = new TestOutputDiagnosticLogger(output);
    }
    
    [Theory]
    [MemberData(nameof(GetData))]
    public async Task Serialization(string name, object target)
    {
        var json = target.ToJsonString(_testOutputLogger);
        await Verify(json).UseParameters(name);
    }
```

**Review by mattjohnsonpint — CHANGES_REQUESTED** [review]

File: test/Sentry.Tests/SerializationTests.cs (hunk 1 above)
Comment (mattjohnsonpint) [reply test/Sentry.Tests/SerializationTests.cs:18]: Specifically, avoid `stream.ToArray()`, and wire up a real logger.

**Review by mattjohnsonpint — APPROVED** [review]

**Comment by SimonCropp** [comment]
from @mattjohnsonpint 

> When I was looking before, I think nint and nuint get automagically treated as IntPtr and UIntPtr

and i confirmed

**Pull request merged**